# Retail Profitability & Inventory Risk Analysis

## Step 1: Import Required Libraries

In this step, we import the Python libraries required for data cleaning, analysis, feature engineering, and basic visualization.

These libraries will help us prepare the retail sales and inventory data for SQL analysis and Tableau dashboarding.


In [14]:
# Import pandas for data loading, cleaning, and transformation
import pandas as pd

# Import numpy for numerical calculations and conditional logic
import numpy as np


## Step 2: Load the Raw Datasets

In this step, we load the original sales and inventory CSV files into Python.

The sales dataset contains transaction-level retail sales data, while the inventory dataset contains stock/on-hand inventory data.

We are loading the files from the `data/raw/` folder because raw data should always remain unchanged in a professional data project.

In [15]:
# Load the raw retail sales dataset
sales = pd.read_csv("retail_sales_ml_apl.csv")

# Load the raw retail inventory dataset
inventory = pd.read_csv("retail_inventory_ml_apl.csv")

In [16]:
# Check the number of rows and columns in each dataset
print("Sales dataset shape:", sales.shape)
print("Inventory dataset shape:", inventory.shape)

Sales dataset shape: (125751, 17)
Inventory dataset shape: (284755, 18)


## Step 3: Preview the Raw Datasets

Before cleaning or transforming the data, we preview the first few rows of both datasets.

This helps us understand the structure of the sales and inventory data, including product fields, store fields, sales values, inventory quantities, and cost-related columns.

In [17]:
# Preview the first 5 rows of the sales dataset
sales.head()

,Transaction Date,Sales Type,Is Return,Reason of Return,Supplier,Product No,Product Description,Product Division,Product Category,Product Subcategory,Product Segment,Store,Sales Channel,Qty Sold,Sales Amount,Cogs,Number of Transactions
0,2025-09-01,Full Price,0,NaN,Vendor 0169,PROD-139867,PROD-100449 SEG-500065 Scholar Footwear Trailb...,Scholar Footwear,Scholar Footwear Trailblazer,Scholar Footwear Trailblazer Equilibrium,SEG-500065 Scholar Footwear Trailblazer Equili...,STR-1006,Channel Alpha,1.0,115.00,52.31,1
1,2025-09-07,Full Price,0,NaN,Vendor 0166,PROD-100449,PROD-100825 SEG-500112 Heritage Footwear Fligh...,Femme Footwear,Femme Footwear Flight Classic,Femme Footwear Flight Classic Varied,SEG-500112 Heritage Footwear Flight Classic Ze...,STR-1198,Channel Alpha,1.0,115.00,58.47,1
2,2025-08-06,Promo,0,NaN,Vendor 0169,PROD-153623,PROD-148813 SEG-511388 Scholar Footwear Trailb...,Scholar Footwear,Scholar Footwear Trailblazer,Scholar Footwear Trailblazer Equilibrium,SEG-511388 Scholar Footwear Trailblazer Equili...,STR-1157,Channel Alpha,1.0,79.99,43.76,1
3,2025-08-31,Full Price,0,NaN,Vendor 0246,PROD-140876,PROD-100419 SEG-500063 Scholar Footwear Boot C...,Scholar Footwear,Scholar Footwear Boot Collection,Scholar Footwear Boot Collection Woodland,SEG-500063 Scholar Footwear Boot Collection Wo...,STR-1234,Channel Alpha,1.0,110.00,50.60,1
4,2026-03-22,Full Price,0,NaN,Vendor 0169,PROD-166113,PROD-149345 SEG-511444 Femme Footwear Trailbla...,Femme Footwear,Femme Footwear Trailblazer,Femme Footwear Trailblazer Equilibrium,SEG-511444 Femme Footwear Trailblazer Equilibrium,STR-1157,Channel Alpha,1.0,160.00,77.95,1


In [18]:
# Preview the first 5 rows of the inventory dataset
inventory.head()

,Start Date,End Date,Stock Status,Supplier,Product No,Product Description,Product Division,Product Category,Product Subcategory,Product Segment,Store,Store Type,Sales Channel,Qty on hand,Stocks Selling Amount,Cost of Stocks,Stock Unit Selling Price,Stock Unit Cost Price
0,2025-08-07,2025-10-18,Full Price,Vendor 0255,PROD-159369,PROD-103395 SEG-500461 Scholar Footwear Canvas...,Scholar Footwear,Scholar Footwear Canvas Styles,Scholar Footwear Canvas Styles Drifter,SEG-500461 Scholar Footwear Canvas Styles Drifter,STR-1010,Retail Storefront,Channel Alpha,1,55.0,24.75,55.0,24.75
1,2026-02-15,2026-02-15,Full Price,Vendor 0010,PROD-180713,PROD-190644 SEG-520887 Femme Footwear Court Ca...,Femme Footwear,Femme Footwear Court Casual,Femme Footwear Court Casual Trident,SEG-520887 Femme Footwear Court Casual Trident,STR-1082,Retail Storefront,Channel Alpha,1,130.0,59.73,130.0,59.73
2,2026-01-08,2026-02-15,Full Price,Vendor 0166,PROD-169474,PROD-153439 SEG-512518 Femme Footwear Court Ca...,Femme Footwear,Femme Footwear Court Casual,Femme Footwear Court Casual Zephyr,SEG-512518 Femme Footwear Court Casual Zephyr,STR-1105,Retail Storefront,Channel Alpha,1,120.0,60.44,120.0,60.44
3,2025-12-23,2026-02-15,Full Price,Vendor 0106,PROD-170976,PROD-122029 SEG-504463 Junior Apparel Cozy Pul...,Junior Apparel,Junior Apparel Cozy Pullovers,Junior Apparel Cozy Pullovers Varied,SEG-504463 Junior Apparel Cozy Pullovers Varied,STR-1157,Retail Storefront,Channel Alpha,3,150.0,52.50,50.0,17.50
4,2026-02-10,9999-12-31,Full Price,Vendor 0134,PROD-183525,PROD-111954 SEG-502325 Scholar Footwear Modern...,Scholar Footwear,Scholar Footwear Modern Legend,Scholar Footwear Modern Legend Summit,SEG-502325 Scholar Footwear Modern Legend Summit,STR-1253,Retail Storefront,Channel Alpha,4,360.0,182.16,90.0,45.54


## Step 4: Check Dataset Information

After previewing the raw data, we check the structure of each dataset using `.info()`.

This helps us understand column names, data types, missing values, and whether date or numeric columns need cleaning before analysis.

In [19]:
# Check column names, non-null counts, and data types for the sales dataset
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125751 entries, 0 to 125750
Data columns (total 17 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   Transaction Date        125751 non-null  object 
 1   Sales Type              125751 non-null  object 
 2   Is Return               125751 non-null  int64  
 3   Reason of Return        0 non-null       float64
 4   Supplier                125751 non-null  object 
 5   Product No              125751 non-null  object 
 6   Product Description     125751 non-null  object 
 7   Product Division        125751 non-null  object 
 8   Product Category        125751 non-null  object 
 9   Product Subcategory     125751 non-null  object 
 10  Product Segment         125751 non-null  object 
 11  Store                   125751 non-null  object 
 12  Sales Channel           125751 non-null  object 
 13  Qty Sold                125751 non-null  float64
 14  Sales Amount        

In [20]:
# Check column names, non-null counts, and data types for the inventory dataset
inventory.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284755 entries, 0 to 284754
Data columns (total 18 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Start Date                284755 non-null  object 
 1   End Date                  284755 non-null  object 
 2   Stock Status              284755 non-null  object 
 3   Supplier                  284755 non-null  object 
 4   Product No                284755 non-null  object 
 5   Product Description       284755 non-null  object 
 6   Product Division          284755 non-null  object 
 7   Product Category          284755 non-null  object 
 8   Product Subcategory       284755 non-null  object 
 9   Product Segment           284755 non-null  object 
 10  Store                     284755 non-null  object 
 11  Store Type                284755 non-null  object 
 12  Sales Channel             284755 non-null  object 
 13  Qty on hand               284755 non-null  i

## Step 5: Standardize Column Names

The raw datasets contain column names with spaces and mixed capitalization.

For a professional analysis workflow, we standardize all column names into lowercase `snake_case`.

This makes the data easier to use in Python, SQL, and Tableau, especially when creating KPIs, joining datasets, and writing SQL queries later.

In [21]:
# Standardize column names in the sales dataset
sales.columns = (sales.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_"))

In [22]:
# Check updated sales column names
sales.columns

Index(['transaction_date', 'sales_type', 'is_return', 'reason_of_return',
       'supplier', 'product_no', 'product_description', 'product_division',
       'product_category', 'product_subcategory', 'product_segment', 'store',
       'sales_channel', 'qty_sold', 'sales_amount', 'cogs',
       'number_of_transactions'],
      dtype='object')

In [23]:
# Standardize column names in the inventory dataset
inventory.columns = (inventory.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_"))

In [24]:
# Check updated inventory column names
inventory.columns

Index(['start_date', 'end_date', 'stock_status', 'supplier', 'product_no',
       'product_description', 'product_division', 'product_category',
       'product_subcategory', 'product_segment', 'store', 'store_type',
       'sales_channel', 'qty_on_hand', 'stocks_selling_amount',
       'cost_of_stocks', 'stock_unit_selling_price', 'stock_unit_cost_price'],
      dtype='object')

## Step 6: Check Missing Values

In this step, we check missing values in the sales and inventory datasets.

Missing values can affect analysis accuracy, especially when calculating revenue, profit, inventory value, stock movement, and risk flags.

Before creating business KPIs, we need to understand whether any important columns have missing data.

In [25]:
# Check missing values in the sales dataset
sales.isnull().sum()

transaction_date               0
sales_type                     0
is_return                      0
reason_of_return          125751
supplier                       0
product_no                     0
product_description            0
product_division               0
product_category               0
product_subcategory            0
product_segment                0
store                          0
sales_channel                  0
qty_sold                       0
sales_amount                   0
cogs                           0
number_of_transactions         0
dtype: int64

In [26]:
# Check missing values in the inventory dataset
inventory.isnull().sum()

start_date                  0
end_date                    0
stock_status                0
supplier                    0
product_no                  0
product_description         0
product_division            0
product_category            0
product_subcategory         0
product_segment             0
store                       0
store_type                  0
sales_channel               0
qty_on_hand                 0
stocks_selling_amount       0
cost_of_stocks              0
stock_unit_selling_price    0
stock_unit_cost_price       0
dtype: int64

## Step 7: Remove Unusable Column

The `reason_of_return` column is completely missing in the sales dataset.

Since all values are empty, this column cannot help explain customer returns or business performance.

We remove it to keep the dataset clean and focused on usable fields.

In [27]:
# Drop the reason_of_return column because it contains only missing values
sales = sales.drop(columns=["reason_of_return"])

In [28]:
# Confirm the column was removed
sales.columns

Index(['transaction_date', 'sales_type', 'is_return', 'supplier', 'product_no',
       'product_description', 'product_division', 'product_category',
       'product_subcategory', 'product_segment', 'store', 'sales_channel',
       'qty_sold', 'sales_amount', 'cogs', 'number_of_transactions'],
      dtype='object')

## Step 8: Convert Date Columns

The sales and inventory datasets contain date columns stored as text.

For proper time-based analysis, these columns need to be converted into datetime format.

This will allow us to analyze monthly sales trends, inventory periods, product movement, and stock risk over time.

In [29]:
# Convert transaction_date in the sales dataset into datetime format
sales["transaction_date"] = pd.to_datetime(sales["transaction_date"])

# Convert start_date in the inventory dataset into datetime format
inventory["start_date"] = pd.to_datetime(inventory["start_date"])

## Step 9: Handle Active Inventory Records

The inventory dataset contains an `end_date` column.

Some records may use `9999-12-31` to represent an active or open inventory record rather than a real end date.

To avoid date conversion errors, we first create a flag to identify active inventory records.

In [30]:
# Create a flag for active inventory records where end_date is marked as 9999-12-31
inventory["is_active_inventory_record"] = np.where(
    inventory["end_date"] == "9999-12-31",
    1,
    0)

In [31]:
# Replace 9999-12-31 with missing value before converting end_date
inventory["end_date_clean"] = inventory["end_date"].replace("9999-12-31", np.nan)

# Convert cleaned end_date into datetime format
inventory["end_date_clean"] = pd.to_datetime(inventory["end_date_clean"])

In [32]:
# Check updated date column data types
sales["transaction_date"].info()

inventory[["start_date", "end_date", "end_date_clean", "is_active_inventory_record"]].info()

<class 'pandas.core.series.Series'>
RangeIndex: 125751 entries, 0 to 125750
Series name: transaction_date
Non-Null Count   Dtype         
--------------   -----         
125751 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 982.6 KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284755 entries, 0 to 284754
Data columns (total 4 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   start_date                  284755 non-null  datetime64[ns]
 1   end_date                    284755 non-null  object        
 2   end_date_clean              262054 non-null  datetime64[ns]
 3   is_active_inventory_record  284755 non-null  int64         
dtypes: datetime64[ns](2), int64(1), object(1)
memory usage: 8.7+ MB


## Step 10: Check Duplicate Rows

Before creating business metrics, we check whether the sales and inventory datasets contain duplicate rows.

Duplicate records can inflate revenue, units sold, inventory quantity, inventory value, and profit calculations.

If duplicates exist, we need to review them before deciding whether to remove them.

In [33]:
# Check duplicate rows in the sales dataset
sales_duplicates = sales.duplicated().sum()

# Check duplicate rows in the inventory dataset
inventory_duplicates = inventory.duplicated().sum()

print("Duplicate rows in sales dataset:", sales_duplicates)
print("Duplicate rows in inventory dataset:", inventory_duplicates)

Duplicate rows in sales dataset: 0
Duplicate rows in inventory dataset: 0


## Step 11: Check Date Ranges

After confirming there are no duplicate rows, we check the date range of the sales and inventory datasets.

This helps us understand the business period covered by the data and confirms whether sales and inventory records can be compared fairly.

For this project, date range validation is important because inventory risk analysis depends on comparing sales demand with stock availability over the same period.

In [34]:
# Check the date range of the sales dataset
print("Sales start date:", sales["transaction_date"].min())
print("Sales end date:", sales["transaction_date"].max())

# Check the start date range of the inventory dataset
print("Inventory start date:", inventory["start_date"].min())
print("Inventory latest start date:", inventory["start_date"].max())

# Check the cleaned end date range of the inventory dataset
print("Inventory end date start:", inventory["end_date_clean"].min())
print("Inventory end date latest:", inventory["end_date_clean"].max())

# Count active inventory records
print("Active inventory records:", inventory["is_active_inventory_record"].sum())

Sales start date: 2025-06-01 00:00:00
Sales end date: 2026-04-24 00:00:00
Inventory start date: 2025-06-01 00:00:00
Inventory latest start date: 2026-04-24 00:00:00
Inventory end date start: 2025-06-01 00:00:00
Inventory end date latest: 2026-04-23 00:00:00
Active inventory records: 22701


## Step 12: Check Sales and Inventory Alignment

Before merging sales and inventory data, we need to confirm whether both datasets share common products and stores.

This is important because the final analysis will compare sales demand with inventory availability.

If products or stores exist in one dataset but not the other, the merge may create missing values or incomplete inventory risk analysis.

For this project, this step validates whether `product_no` and `store` can be used as reliable join keys.

In [35]:
# Check product and store counts in both datasets
print("Unique products in sales:", sales["product_no"].nunique())
print("Unique products in inventory:", inventory["product_no"].nunique())

print("Unique stores in sales:", sales["store"].nunique())
print("Unique stores in inventory:", inventory["store"].nunique())

Unique products in sales: 2326
Unique products in inventory: 2326
Unique stores in sales: 40
Unique stores in inventory: 40


In [36]:
# Compare product coverage between sales and inventory
sales_products = set(sales["product_no"].unique())
inventory_products = set(inventory["product_no"].unique())

products_only_in_sales = sales_products - inventory_products
products_only_in_inventory = inventory_products - sales_products

print("Products only in sales:", len(products_only_in_sales))
print("Products only in inventory:", len(products_only_in_inventory))

Products only in sales: 0
Products only in inventory: 0


In [37]:
# Compare store coverage between sales and inventory
sales_stores = set(sales["store"].unique())
inventory_stores = set(inventory["store"].unique())

stores_only_in_sales = sales_stores - inventory_stores
stores_only_in_inventory = inventory_stores - sales_stores

print("Stores only in sales:", len(stores_only_in_sales))
print("Stores only in inventory:", len(stores_only_in_inventory))

Stores only in sales: 0
Stores only in inventory: 0


### Sales and Inventory Alignment Finding

The sales and inventory datasets contain matching product and store coverage.

No products or stores were found only in one dataset.

This means the datasets are well aligned and can be merged later using `product_no` and `store` as key fields.

## Step 13: Review Numeric Summary Statistics

Before creating business KPIs, we review the summary statistics of numeric columns in both datasets.

This helps us quickly identify the minimum, maximum, average, and spread of values such as quantity, sales amount, cost, inventory quantity, and stock value.

This step helps detect negative values, zero values, unusual values, and possible business events such as returns or inventory adjustments.

In [38]:
# Review summary statistics for all numeric columns in the sales dataset
sales.describe()

,transaction_date,is_return,qty_sold,sales_amount,cogs,number_of_transactions
count,125751,125751.00000,125751.000000,125751.000000,125751.000000,125751.000000
mean,2025-11-12 05:16:43.039339264,0.05957,0.930370,83.356079,46.502621,1.042695
min,2025-06-01 00:00:00,0.00000,-3.000000,-380.000000,-192.280000,1.000000
25%,2025-08-24 00:00:00,0.00000,1.000000,50.000000,37.710000,1.000000
50%,2025-11-21 00:00:00,0.00000,1.000000,90.000000,50.050000,1.000000
75%,2026-01-29 00:00:00,0.00000,1.000000,115.000000,60.720000,1.000000
max,2026-04-24 00:00:00,1.00000,14.000000,2660.000000,1345.960000,14.000000
std,NaN,0.23669,0.586897,83.658776,43.597511,0.305832


In [39]:
# Review summary statistics for all numeric columns in the inventory dataset
inventory.describe()

,start_date,qty_on_hand,stocks_selling_amount,cost_of_stocks,stock_unit_selling_price,stock_unit_cost_price,is_active_inventory_record,end_date_clean
count,284755,284755.000000,284755.000000,284755.000000,284755.000000,284755.000000,284755.000000,262054
mean,2025-10-27 03:15:02.253515776,2.294629,224.392566,111.046256,95.313171,47.419488,0.079721,2025-11-15 07:10:13.110274816
min,2025-06-01 00:00:00,-9.000000,-828.000000,-418.950000,0.000000,-16.940000,0.000000,2025-06-01 00:00:00
25%,2025-08-12 00:00:00,1.000000,95.000000,46.430000,75.000000,37.950000,0.000000,2025-09-01 00:00:00
50%,2025-10-25 00:00:00,2.000000,160.000000,80.960000,100.000000,50.050000,0.000000,2025-11-17 00:00:00
75%,2026-01-06 00:00:00,3.000000,300.000000,149.170000,120.000000,60.720000,0.000000,2026-01-21 00:00:00
max,2026-04-24 00:00:00,56.000000,5600.000000,2641.450000,250.000000,336.380000,1.000000,2026-04-23 00:00:00
std,NaN,2.007395,214.479811,105.785092,46.655771,23.776569,0.270862,NaN


## Step 14: Investigate Negative Values

The summary statistics show negative values in sales and inventory.

Before creating KPIs, we check whether these negatives are valid business events such as returns or inventory adjustments.

In [40]:
# Count negative values in sales columns
print("Negative qty_sold:", (sales["qty_sold"] < 0).sum())
print("Negative sales_amount:", (sales["sales_amount"] < 0).sum())
print("Negative cogs:", (sales["cogs"] < 0).sum())

Negative qty_sold: 7491
Negative sales_amount: 7491
Negative cogs: 7491


In [41]:
# Count negative values in inventory columns
print("Negative qty_on_hand:", (inventory["qty_on_hand"] < 0).sum())
print("Negative stocks_selling_amount:", (inventory["stocks_selling_amount"] < 0).sum())
print("Negative cost_of_stocks:", (inventory["cost_of_stocks"] < 0).sum())
print("Negative stock_unit_cost_price:", (inventory["stock_unit_cost_price"] < 0).sum())

Negative qty_on_hand: 1380
Negative stocks_selling_amount: 1380
Negative cost_of_stocks: 8022
Negative stock_unit_cost_price: 25


In [42]:
# Check if negative sales are linked to returns
sales.groupby("is_return")[["qty_sold", "sales_amount", "cogs"]].min()

,qty_sold,sales_amount,cogs
is_return,,,
0,0.0,0.0,0.00
1,-3.0,-380.0,-192.28


## Step 15: Flag Inventory Data Issues

Sales negative values are linked to product returns, so we keep those records.

Inventory negative values may affect stock and inventory value analysis.

Instead of deleting these records, we create a simple flag to identify inventory rows with possible data issues.

In [43]:
# Flag inventory records with negative stock quantity or negative stock value
inventory["inventory_data_issue_flag"] = np.where(
    (inventory["qty_on_hand"] < 0) |
    (inventory["stocks_selling_amount"] < 0) |
    (inventory["cost_of_stocks"] < 0),
    1,
    0)

# Count inventory records with possible data issues
inventory["inventory_data_issue_flag"].sum()

8026

### Inventory Data Issue Finding

The inventory dataset contains 8,026 rows with negative stock quantity or negative stock value.

These records are not removed at this stage. They are flagged using `inventory_data_issue_flag` so they can be reviewed or filtered later during inventory risk analysis.

## Step 16: Create Monthly Sales Features

We create monthly date features from the sales transaction date.

These columns will help analyze sales trends, product movement, and inventory demand by month.

In [44]:
# Create monthly sales features
sales["year_month"] = sales["transaction_date"].dt.to_period("M").astype(str)
sales["month"] = sales["transaction_date"].dt.month

# Preview new monthly features
sales[["transaction_date", "year_month", "month"]].head()

,transaction_date,year_month,month
0,2025-09-01,2025-09,9
1,2025-09-07,2025-09,9
2,2025-08-06,2025-08,8
3,2025-08-31,2025-08,8
4,2026-03-22,2026-03,3


## Step 17: Create Sales Business Metrics

We create sales metrics needed for profitability analysis.

These metrics separate normal sales from returns and calculate gross profit for each transaction.

In [45]:
# Create gross sales amount for non-return transactions
sales["gross_sales_amount"] = np.where(
    sales["is_return"] == 0,
    sales["sales_amount"],
    0)

# Create return amount as a positive value
sales["return_amount"] = np.where(
    sales["is_return"] == 1,
    sales["sales_amount"].abs(),
    0)

# Keep net sales amount including returns
sales["net_sales_amount"] = sales["sales_amount"]

# Create gross profit
sales["gross_profit"] = sales["sales_amount"] - sales["cogs"]

In [46]:
# Preview new sales metrics
sales[["is_return",
        "qty_sold",
        "sales_amount",
        "cogs",
        "gross_sales_amount",
        "return_amount",
        "net_sales_amount",
        "gross_profit"]].head()

,is_return,qty_sold,sales_amount,cogs,gross_sales_amount,return_amount,net_sales_amount,gross_profit
0,0,1.0,115.00,52.31,115.00,0.0,115.00,62.69
1,0,1.0,115.00,58.47,115.00,0.0,115.00,56.53
2,0,1.0,79.99,43.76,79.99,0.0,79.99,36.23
3,0,1.0,110.00,50.60,110.00,0.0,110.00,59.40
4,0,1.0,160.00,77.95,160.00,0.0,160.00,82.05


## Step 18: Create Inventory Value Metrics

We create inventory value metrics needed to measure cash tied up in stock.

These fields will help identify products with high inventory value, overstock risk, and weak sales movement.

In [47]:
# Create inventory value metrics
inventory["inventory_value_at_cost"] = inventory["cost_of_stocks"]
inventory["inventory_value_at_retail"] = inventory["stocks_selling_amount"]

In [48]:
# Preview inventory value metrics
inventory[["qty_on_hand",
        "cost_of_stocks",
        "stocks_selling_amount",
        "inventory_value_at_cost",
        "inventory_value_at_retail"]].head()

,qty_on_hand,cost_of_stocks,stocks_selling_amount,inventory_value_at_cost,inventory_value_at_retail
0,1,24.75,55.0,24.75,55.0
1,1,59.73,130.0,59.73,130.0
2,1,60.44,120.0,60.44,120.0
3,3,52.50,150.0,52.50,150.0
4,4,182.16,360.0,182.16,360.0


## Step 19: Create Current Inventory Dataset

We filter the inventory data to active stock records.

This is necessary because the project focuses on current inventory value, cash tied up in stock, profitability,

overstock risk, and products with high stock but weak sales. 

In [49]:
# Keep only active inventory records
current_inventory = inventory[inventory["is_active_inventory_record"] == 1].copy()

# Check current inventory size
current_inventory.shape

(22701, 23)

In [50]:
# Preview current inventory data
current_inventory.head()

,start_date,end_date,stock_status,supplier,product_no,product_description,product_division,product_category,product_subcategory,product_segment,...,qty_on_hand,stocks_selling_amount,cost_of_stocks,stock_unit_selling_price,stock_unit_cost_price,is_active_inventory_record,end_date_clean,inventory_data_issue_flag,inventory_value_at_cost,inventory_value_at_retail
4,2026-02-10,9999-12-31,Full Price,Vendor 0134,PROD-183525,PROD-111954 SEG-502325 Scholar Footwear Modern...,Scholar Footwear,Scholar Footwear Modern Legend,Scholar Footwear Modern Legend Summit,SEG-502325 Scholar Footwear Modern Legend Summit,...,4,360.0,182.16,90.0,45.540,1,NaT,0,182.16,360.0
30,2025-10-07,9999-12-31,Full Price,Vendor 0262,PROD-176445,PROD-178967 SEG-518011 Femme Footwear Trailbla...,Femme Footwear,Femme Footwear Trailblazer,Femme Footwear Trailblazer Varied,SEG-518011 Femme Footwear Trailblazer Varied,...,1,150.0,50.00,150.0,50.000,1,NaT,0,50.00,150.0
35,2025-11-17,9999-12-31,Full Price,Vendor 0169,PROD-166155,PROD-149165 SEG-511428 Femme Footwear Trailbla...,Femme Footwear,Femme Footwear Trailblazer,Femme Footwear Trailblazer Equilibrium,SEG-511428 Femme Footwear Trailblazer Equilibrium,...,2,230.0,115.13,115.0,57.565,1,NaT,0,115.13,230.0
36,2025-11-17,9999-12-31,Full Price,Vendor 0169,PROD-166156,PROD-149166 SEG-511428 Femme Footwear Trailbla...,Femme Footwear,Femme Footwear Trailblazer,Femme Footwear Trailblazer Equilibrium,SEG-511428 Femme Footwear Trailblazer Equilibrium,...,2,230.0,115.12,115.0,57.560,1,NaT,0,115.12,230.0
37,2025-11-17,9999-12-31,Full Price,Vendor 0169,PROD-166157,PROD-149167 SEG-511428 Femme Footwear Trailbla...,Femme Footwear,Femme Footwear Trailblazer,Femme Footwear Trailblazer Equilibrium,SEG-511428 Femme Footwear Trailblazer Equilibrium,...,2,230.0,115.12,115.0,57.560,1,NaT,0,115.12,230.0


## Step 20: Aggregate Sales by Product and Store

We summarize sales at the product-store level.

This is necessary because the inventory data is also analyzed by product and store. 

Creating this sales summary allows us to later compare sales performance with current stock availability

In [51]:
# Aggregate sales at product-store level
sales_summary = sales.groupby(["product_no",
        "store",
        "supplier",
        "product_description",
        "product_division",
        "product_category",
        "product_subcategory",
        "product_segment",
        "sales_channel"],
    as_index=False).agg(
    total_gross_sales=("gross_sales_amount", "sum"),
    total_returns=("return_amount", "sum"),
    total_net_sales=("net_sales_amount", "sum"),
    total_cogs=("cogs", "sum"),
    total_gross_profit=("gross_profit", "sum"),
    total_qty_sold=("qty_sold", "sum"),
    total_transactions=("number_of_transactions", "sum"))

In [52]:
# Preview sales summary
sales_summary.head()

,product_no,store,supplier,product_description,product_division,product_category,product_subcategory,product_segment,sales_channel,total_gross_sales,total_returns,total_net_sales,total_cogs,total_gross_profit,total_qty_sold,total_transactions
0,PROD-100043,STR-1006,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,96.0,0.0,96.0,68.74,27.26,1.0,1
1,PROD-100043,STR-1010,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,0.0,160.0,68.02,91.98,1.0,1
2,PROD-100043,STR-1026,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,160.0,0.0,0.00,0.00,0.0,2
3,PROD-100043,STR-1074,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,112.0,112.0,0.0,0.01,-0.01,0.0,2
4,PROD-100043,STR-1149,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,160.0,0.0,0.00,0.00,0.0,2


## Step 21: Aggregate Current Inventory by Product and Store

We summarize current inventory at the product-store level.

This prepares the inventory data to be merged with the sales summary,

so we can compare sales performance with current stock availability and inventory value.

In [53]:
# Aggregate current inventory at product-store level
inventory_summary = current_inventory.groupby(["product_no","store"],as_index=False).agg(
    current_qty_on_hand=("qty_on_hand", "sum"),
    current_inventory_value_at_cost=("inventory_value_at_cost", "sum"),
    current_inventory_value_at_retail=("inventory_value_at_retail", "sum"),
    inventory_data_issue_flag=("inventory_data_issue_flag", "max"))

In [54]:
# Preview inventory summary
inventory_summary.head()

,product_no,store,current_qty_on_hand,current_inventory_value_at_cost,current_inventory_value_at_retail,inventory_data_issue_flag
0,PROD-100043,STR-1006,1,67.61,160.0,0
1,PROD-100043,STR-1010,2,136.03,320.0,0
2,PROD-100043,STR-1026,1,69.01,160.0,0
3,PROD-100043,STR-1044,1,69.14,160.0,0
4,PROD-100043,STR-1074,1,67.99,160.0,0


## Step 22: Merge Sales and Current Inventory

We merge product-store sales performance with current inventory.

This creates one analysis dataset that can be used to identify profitable products, overstock risk, stockout risk, and cash tied up in inventory.

In [55]:
# Merge sales summary with current inventory summary
analysis_df = sales_summary.merge(inventory_summary,
    on=["product_no", "store"],
    how="left")

In [56]:
# Preview merged dataset
analysis_df.head()

,product_no,store,supplier,product_description,product_division,product_category,product_subcategory,product_segment,sales_channel,total_gross_sales,total_returns,total_net_sales,total_cogs,total_gross_profit,total_qty_sold,total_transactions,current_qty_on_hand,current_inventory_value_at_cost,current_inventory_value_at_retail,inventory_data_issue_flag
0,PROD-100043,STR-1006,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,96.0,0.0,96.0,68.74,27.26,1.0,1,1.0,67.61,160.0,0.0
1,PROD-100043,STR-1010,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,0.0,160.0,68.02,91.98,1.0,1,2.0,136.03,320.0,0.0
2,PROD-100043,STR-1026,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,160.0,0.0,0.00,0.00,0.0,2,1.0,69.01,160.0,0.0
3,PROD-100043,STR-1074,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,112.0,112.0,0.0,0.01,-0.01,0.0,2,1.0,67.99,160.0,0.0
4,PROD-100043,STR-1149,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,160.0,0.0,0.00,0.00,0.0,2,NaN,NaN,NaN,NaN


In [57]:
# Check merged dataset size
analysis_df.shape

(50447, 20)

## Step 23: Check Missing Values After Merge

After merging sales with current inventory, we check missing values in the full analysis dataset.

This shows whether any fields became missing because some product-store sales records did not have matching current 

inventory records.

In [58]:
# Check missing values in the merged analysis dataset
analysis_df.isnull().sum()

product_no                               0
store                                    0
supplier                                 0
product_description                      0
product_division                         0
product_category                         0
product_subcategory                      0
product_segment                          0
sales_channel                            0
total_gross_sales                        0
total_returns                            0
total_net_sales                          0
total_cogs                               0
total_gross_profit                       0
total_qty_sold                           0
total_transactions                       0
current_qty_on_hand                  35760
current_inventory_value_at_cost      35760
current_inventory_value_at_retail    35760
inventory_data_issue_flag            35760
dtype: int64

## Step 24: Handle Missing Inventory Values

Some product-store records have sales history but no matching current inventory record.

For inventory fields, missing values mean there is no active stock record available. 

We replace these missing inventory values with 0 so inventory risk metrics can be calculated correctly.

In [59]:
# Replace missing inventory values with 0
analysis_df = analysis_df.fillna({"current_qty_on_hand": 0,
    "current_inventory_value_at_cost": 0,
    "current_inventory_value_at_retail": 0,
    "inventory_data_issue_flag": 0})

In [60]:
# Confirm missing values are handled
analysis_df.isnull().sum()

product_no                           0
store                                0
supplier                             0
product_description                  0
product_division                     0
product_category                     0
product_subcategory                  0
product_segment                      0
sales_channel                        0
total_gross_sales                    0
total_returns                        0
total_net_sales                      0
total_cogs                           0
total_gross_profit                   0
total_qty_sold                       0
total_transactions                   0
current_qty_on_hand                  0
current_inventory_value_at_cost      0
current_inventory_value_at_retail    0
inventory_data_issue_flag            0
dtype: int64

## Step 25: Create Profitability Metrics

We calculate gross margin percentage to measure profitability.

This helps compare products and stores based on profit efficiency, not only sales revenue.

In [61]:
# Calculate gross margin percentage
analysis_df["gross_margin_pct"] = np.where(analysis_df["total_net_sales"] > 0,
    analysis_df["total_gross_profit"] / analysis_df["total_net_sales"],0)

In [62]:
# Preview profitability metric
analysis_df.head()

,product_no,store,supplier,product_description,product_division,product_category,product_subcategory,product_segment,sales_channel,total_gross_sales,...,total_net_sales,total_cogs,total_gross_profit,total_qty_sold,total_transactions,current_qty_on_hand,current_inventory_value_at_cost,current_inventory_value_at_retail,inventory_data_issue_flag,gross_margin_pct
0,PROD-100043,STR-1006,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,96.0,...,96.0,68.74,27.26,1.0,1,1.0,67.61,160.0,0.0,0.283958
1,PROD-100043,STR-1010,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,...,160.0,68.02,91.98,1.0,1,2.0,136.03,320.0,0.0,0.574875
2,PROD-100043,STR-1026,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,...,0.0,0.00,0.00,0.0,2,1.0,69.01,160.0,0.0,0.000000
3,PROD-100043,STR-1074,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,112.0,...,0.0,0.01,-0.01,0.0,2,1.0,67.99,160.0,0.0,0.000000
4,PROD-100043,STR-1149,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,...,0.0,0.00,0.00,0.0,2,0.0,0.00,0.0,0.0,0.000000


## Step 26: Create Stock-to-Sales Ratio

We calculate the stock-to-sales ratio to compare current stock with sales demand.

This helps identify products with high stock but weak sales, which may indicate overstock risk or cash tied up in inventory.

In [63]:
# Calculate stock-to-sales ratio
analysis_df["stock_to_sales_ratio"] = np.where(
    analysis_df["total_qty_sold"] > 0,
    analysis_df["current_qty_on_hand"] / analysis_df["total_qty_sold"],
    0)

In [64]:
# Preview updated analysis dataset
analysis_df.head()

,product_no,store,supplier,product_description,product_division,product_category,product_subcategory,product_segment,sales_channel,total_gross_sales,...,total_cogs,total_gross_profit,total_qty_sold,total_transactions,current_qty_on_hand,current_inventory_value_at_cost,current_inventory_value_at_retail,inventory_data_issue_flag,gross_margin_pct,stock_to_sales_ratio
0,PROD-100043,STR-1006,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,96.0,...,68.74,27.26,1.0,1,1.0,67.61,160.0,0.0,0.283958,1.0
1,PROD-100043,STR-1010,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,...,68.02,91.98,1.0,1,2.0,136.03,320.0,0.0,0.574875,2.0
2,PROD-100043,STR-1026,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,...,0.00,0.00,0.0,2,1.0,69.01,160.0,0.0,0.000000,0.0
3,PROD-100043,STR-1074,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,112.0,...,0.01,-0.01,0.0,2,1.0,67.99,160.0,0.0,0.000000,0.0
4,PROD-100043,STR-1149,Vendor 0064,PROD-189472 SEG-520541 Femme Footwear Boot Col...,Femme Footwear,Femme Footwear Boot Collection,Femme Footwear Boot Collection Plushfoot,SEG-520541 Femme Footwear Boot Collection Plus...,Channel Alpha,160.0,...,0.00,0.00,0.0,2,0.0,0.00,0.0,0.0,0.000000,0.0


## Step 27: Create Inventory Risk Flags

We create simple inventory risk flags using sales demand and current stock.

These flags help identify fast-moving products, slow-moving products, stockout risk, and overstock risk.

In [65]:
# Calculate sales and stock thresholds for risk flags
high_sales_threshold = analysis_df["total_qty_sold"].quantile(0.75)
low_sales_threshold = analysis_df["total_qty_sold"].quantile(0.25)
high_stock_threshold = analysis_df["current_qty_on_hand"].quantile(0.75)

In [66]:
# Fast-moving flag:
# Product-store combinations with high quantity sold.
# These items may need frequent replenishment.
analysis_df["fast_moving_flag"] = np.where(
    analysis_df["total_qty_sold"] >= high_sales_threshold,
    1,0)

In [67]:
# Slow-moving flag:
# Product-store combinations with low quantity sold.
# These items may need promotion, markdown, or review.
analysis_df["slow_moving_flag"] = np.where(
    analysis_df["total_qty_sold"] <= low_sales_threshold,
    1,0)

In [68]:
# Stockout risk flag:
# Product-store combinations with strong sales but no current stock.
# These items may need restocking.
analysis_df["stockout_risk_flag"] = np.where(
    (analysis_df["total_qty_sold"] >= high_sales_threshold) &
    (analysis_df["current_qty_on_hand"] == 0),
    1,0)

In [69]:
# Overstock risk flag:
# Product-store combinations with weak sales but high current stock.
# These items may be tying up inventory cash.
analysis_df["overstock_risk_flag"] = np.where(
    (analysis_df["total_qty_sold"] <= low_sales_threshold) &
    (analysis_df["current_qty_on_hand"] >= high_stock_threshold),
    1,
    0)

In [70]:
# Check number of records flagged in each risk category
analysis_df[["fast_moving_flag",
        "slow_moving_flag",
        "stockout_risk_flag",
        "overstock_risk_flag"]].sum()

fast_moving_flag       13997
slow_moving_flag       21658
stockout_risk_flag     10323
overstock_risk_flag    21585
dtype: int64

## Step 28: Final Dataset Check Before Export

Before exporting the cleaned analysis dataset, we check its structure, missing values, and final shape.

This confirms that the dataset is ready for SQL analysis and Tableau dashboarding.

In [71]:
# Check final dataset shape and columns
print("Final analysis dataset shape:", analysis_df.shape)

analysis_df.info()

Final analysis dataset shape: (50447, 26)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50447 entries, 0 to 50446
Data columns (total 26 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   product_no                         50447 non-null  object 
 1   store                              50447 non-null  object 
 2   supplier                           50447 non-null  object 
 3   product_description                50447 non-null  object 
 4   product_division                   50447 non-null  object 
 5   product_category                   50447 non-null  object 
 6   product_subcategory                50447 non-null  object 
 7   product_segment                    50447 non-null  object 
 8   sales_channel                      50447 non-null  object 
 9   total_gross_sales                  50447 non-null  float64
 10  total_returns                      50447 non-null  float64
 11  total_net_sa

In [72]:
# Check final missing values
analysis_df.isnull().sum()

product_no                           0
store                                0
supplier                             0
product_description                  0
product_division                     0
product_category                     0
product_subcategory                  0
product_segment                      0
sales_channel                        0
total_gross_sales                    0
total_returns                        0
total_net_sales                      0
total_cogs                           0
total_gross_profit                   0
total_qty_sold                       0
total_transactions                   0
current_qty_on_hand                  0
current_inventory_value_at_cost      0
current_inventory_value_at_retail    0
inventory_data_issue_flag            0
gross_margin_pct                     0
stock_to_sales_ratio                 0
fast_moving_flag                     0
slow_moving_flag                     0
stockout_risk_flag                   0
overstock_risk_flag      

## Step 29: Export Cleaned Datasets

We export the cleaned datasets for the next stages of the project.

The final analysis dataset will be used for SQL business analysis and Tableau dashboarding.

In [73]:
# Export cleaned sales data
sales.to_csv("cleaned_sales.csv", index=False)

# Export cleaned inventory data
inventory.to_csv("cleaned_inventory.csv", index=False)

# Export final analysis dataset
analysis_df.to_csv("retail_analysis_dataset.csv", index=False)

## Final Summary

The Retail Profitability and Inventory Risk dataset was cleaned, validated, transformed, and prepared for business analysis.

In this notebook, the raw sales and inventory datasets were processed to create a final analysis-ready dataset at the product-store level. The cleaning process included standardizing column names, checking missing values, removing unusable columns, converting date fields, validating duplicate records, reviewing date ranges, and checking alignment between sales and inventory data.

Several business metrics were created to support profitability and inventory analysis, including gross sales, return amount, net sales, gross profit, current inventory value at cost, current inventory value at retail, gross margin percentage, and stock-to-sales ratio.

Inventory data quality issues were not removed directly. Instead, negative inventory-related values were flagged using an `inventory_data_issue_flag` so that data issues can be reviewed without losing important records.

The final dataset also includes inventory risk indicators such as `fast_moving_flag`, `slow_moving_flag`, `stockout_risk_flag`, and `overstock_risk_flag`. These flags help identify products that may need restocking, promotion, markdown, or inventory review.

Final outputs created:
- `cleaned_sales.csv`
- `cleaned_inventory.csv`
- `retail_analysis_dataset.csv`

The final analysis dataset contains 50,447 rows and 26 columns with no missing values. It is ready for SQL analysis, Tableau dashboard creation, and GitHub portfolio presentation.